Tests/NIOTests/DatagramChannelTests.swift

```diff
@@ -662,4 +666,85 @@ final class DatagramChannelTests: XCTestCase {
             XCTAssertFalse(try channel2.getOption(ChannelOptions.explicitCongestionNotification).wait())
         } ())
     }
+    
+    private func testEcnReceive(address: String, vectorRead: Bool, vectorSend: Bool) {
+        XCTAssertNoThrow(try {
+            let receiveBootstrap: DatagramBootstrap
+            if vectorRead {
+                receiveBootstrap = DatagramBootstrap(group: group)
+                    .channelOption(ChannelOptions.datagramVectorReadMessageCount, value: 4)
+            } else {
+                receiveBootstrap = DatagramBootstrap(group: group)
+            }
+                
+            let receiveChannel = try receiveBootstrap
+                .channelOption(ChannelOptions.explicitCongestionNotification, value: true)
+                .channelInitializer { channel in
+                    channel.pipeline.addHandler(DatagramReadRecorder<ByteBuffer>(), name: "ByteReadRecorder")
+                }
+                .bind(host: address, port: 0)
+                .wait()
+            let sendChannel = try DatagramBootstrap(group: group)
+                .bind(host: address, port: 0)
+                .wait()
+            
+            var buffer = sendChannel.allocator.buffer(capacity: 1)
+            buffer.writeRepeatingByte(0, count: 1)
+            let ecnStates: [NIOExplicitCongestionNotificationState] = [.transportNotCapable,
+                                                                       .congestionExperienced,
+                                                                       .transportCapableFlag0,
+                                                                       .transportCapableFlag1]
+            for ecnState in ecnStates {
+                let writeData = AddressedEnvelope(remoteAddress: receiveChannel.localAddress!,
+                                                  data: buffer,
+                                                  metadata: .init(ecnState: ecnState))
+                // Sending extra data without flushing should trigger a vector send.
+                if (vectorSend) {
+                    sendChannel.write(writeData, promise: nil)
+                }
+                try sendChannel.writeAndFlush(writeData).wait()
+            }
+
+            let expectedReads = ecnStates.count * (vectorSend ? 2 : 1)
+            let reads = try receiveChannel.waitForDatagrams(count: expectedReads)
+            XCTAssertEqual(reads.count, expectedReads)
+            for readNumber in 0..<reads.count {
+                let read = reads[readNumber]
+                XCTAssertEqual(read.metadata?.ecnState, ecnStates[readNumber / (vectorSend ? 2 : 1)])
+            }
+        } ())
```

Comment:
This test leaks sockets: can we make sure we're adding `defer` blocks to clean up channels?

Reply:
Done.